asserts/snapasserts/validation_sets.go

```diff
@@ -44,6 +60,62 @@ func (e *ValidationSetsConflictError) Error() string {
 	return buf.String()
 }
 
+// ValidationSetsValidationError describes an error arising
+// from validation of snaps against ValidationSets.
+type ValidationSetsValidationError struct {
+	// MissingSnaps maps missing snap names to the validation sets requiring them.
+	MissingSnaps map[string][]string
+	// InvalidSnaps maps snap names to the validation sets declaring them invalid.
+	InvalidSnaps map[string][]string
+	// WronRevisionSnaps maps snap names to the expected revisions and respective
+	// validation sets that require them.
+	WrongRevisionSnaps map[string]map[snap.Revision][]string
+	// Sets maps validation set keys referenced by above maps to actual
+	// validation sets.
+	Sets map[string]*asserts.ValidationSet
+}
+
+func (e *ValidationSetsValidationError) Error() string {
+	buf := bytes.NewBufferString("validation sets assertions are not met:")
+	printDetails := func(header string, details map[string][]string,
+		printSnap func(snapName string, keys []string) string) {
+		if len(details) == 0 {
+			return
+		}
+		fmt.Fprintf(buf, "\n- %s:", header)
+		for snapName, validationSetKeys := range details {
+			fmt.Fprintf(buf, "\n  - %s", printSnap(snapName, validationSetKeys))
+		}
+	}
+
+	printDetails("missing required snaps", e.MissingSnaps, func(snapName string, validationSetKeys []string) string {
+		return fmt.Sprintf("%s (required by sets %s)", snapName, strings.Join(validationSetKeys, ","))
+	})
+	printDetails("invalid snaps", e.InvalidSnaps, func(snapName string, validationSetKeys []string) string {
+		return fmt.Sprintf("%s (invalid for sets %s)", snapName, strings.Join(validationSetKeys, ","))
```

Comment:
I checked it with @degville, quoting him: "depends slightly on the context. If a, b or c are a collection, for example, and foo fails against only one element then it might be 'in'. but 'for' is a safe bet as it works both ways."

So, leaving it as is at "for" as this error means the snap is invalid in every listed set.